Proposal for improving SDK selection with the host and admin installs

marcpopMSFT opened the pull request:
Proposal for improving the SDK selection for the host and have it work better with admin installs (especially visual Studio).

Include a proposal for fast SDK version switching. Not-included is api work to share the SDK location to callers of the host/dotnetup.

marcpopMSFT committed:
Created a proposal for an SDK path environment variable for improving the SDK selection for the host and have it work better with admin installs (especially visual Studio).
Include a proposal for fast SDK version switching.
Not-included is api work to share the SDK location to callers of the host/dotnetup.

jkotas commented:
If I understand this correctly, this environment variable is meant to allow de-coupling location of `dotnet.exe`/`dotnet` and the actual SDK bits.

Current state: The SDK has to be in the known directory structure under that `dotnet` used to launch it. For example, if dotnet is located at `/foo/dotnet`, the SDK has to be located at `/foo`  (/foo/sdk/10.0.123/,..) too.

Proposal change: Optionally override the SDK location using env variable. For example, if dotnet is located at `/foo/dotnet`, the SDK can be located at `/bar` (e.g. /bar/sdk/10.0.123/,..). 

Is this correct understanding?

baronfel commented:
That's a huge part of it, yeah. The thought being that if we can't always control PATH on Windows due to the choices we've made in our MSIs in the past, we can at least make the dotnet binaries that are invoked use the SDKs _and_ Runtimes that the user intended.

Then as a value-add on top of that, if the mixer _also_ allowed for specific SDK version selection to be driven externally then you can put together a really good UX around experimenting with different SDK versions in the same workspace without having to make any permanent changes. This is something that folks struggle with today, as you have to create/update global.jsons to get the same experience. This isn't _high_ friction, but it's also not _no_ friction, and there's substantial prior art in other ecosystems for making it easy to switch tool chains on a temporary basis.

jkotas commented:
> we can at least make the dotnet binaries that are invoked use the SDKs and Runtimes that the user intended.

Won't you also need this to control the runtime selection for `dotnet myapp.dll` to make it work well? The proposal talks about SDK selection only.

jjonescz commented on accepted/2025/sdk-selection-improvements.md:
```diff
@@ -0,0 +1,176 @@
+# **Improving .NET SDK Selection**
```
nit: having headings `**bolded**` seems superfluous

marcpopMSFT commented:
> > we can at least make the dotnet binaries that are invoked use the SDKs and Runtimes that the user intended.
> 
> Won't you also need this to control the runtime selection for `dotnet myapp.dll` to make it work well? The proposal talks about SDK selection only.

For the dotnetup scenarios, we are going to be setting DOTNET_ROOT in addition to DOTNET_SDK_ROOT. For the muxer, we are proposing that the environment variable function like the global.json paths setting (with the addition of env expansion). I previously suggested having the muxer override DOTNET_ROOT for the global.json path scenario but there was pushback hence keeping two separate values.

What's the behavior today if someone uses global.json paths and calls `dotnet myapp.dll`?

marcpopMSFT commented on accepted/2025/sdk-selection-improvements.md:
```diff
@@ -0,0 +1,176 @@
+*   Set both:
+    *   `DOTNET_SDK_ROOT`
+    *   `DOTNET_ROOT`
+*   Modify the **PATH** in the current window to point to the selected SDK.
```
Isn't the reason to reset the PATH to the same location as DOTNET_SDK_ROOT specifically to handle the case where the current path points to a <11 host?

nagilson replied:
I still think we should set the PATH, but this might have unintended side effects. Here's what I mean:

If I run `dotnetup use` -> that uses the `dotnetup` hive. The user does not have .NET 11 in the dotnetup hive. Either they never installed 11, or they uninstalled 11. 

Then:
1. The dotnet.exe that's on the PATH of the terminal will not respect or be aware of DOTNET_SDK_ROOT
2. The dotnet.exe that's on the PATH will still correctly find what's in the local `dotnetup` hive since that's what is alongside of it
3. If we ever rely on DOTNET_SDK_ROOT to help find a custom directory install that is not in the hive, then that would not work
4. If the user set DOTNET_SDK_ROOT to something other than the `dotnetup` hive,  then ran `dotnetup`, DOTNET_SDK_ROOT would no longer work
5. If the DOTNET_HOST_PATH or other host logic is required to ensure subprocesses have the correct path of the SDK/Host that is used, then that will not run, which could cause issues.

jkotas commented:
> What's the behavior today if someone uses global.json paths and calls dotnet myapp.dll?

I believe that `dotnet myapp.dll` does not consult global.json nor DOTNET_ROOT. `dotnet myapp.dll` is going to run on the runtime that is in the subdirectory of where `dotnet` is located.

nagilson commented on accepted/2025/sdk-selection-improvements.md:
```diff
@@ -0,0 +1,176 @@
+*   **User Experience:**
+    *   Detect if the global path points to `Program Files/dotnet` and the customer doesn't have the .NET 11 host which supports the SDK root feature.
+    *   Notify the user:
+        > This configuration will only work in this window. Use `dotnetup use` each time you open a new command prompt.
```
Right - I'd have to open my IDE/workspace in a way that gets the effects of running dotnetup (aka launch from the terminal). But users might not notice that nuance.

jkotas commented:
> So in the case of someone using dotnetup where we've set both DOTNET_SDK_ROOT and DOTNET_ROOT, they'll get the runtime they expect.

They will get the global runtime, they won't get the runtime that `DOTNET_SDK_ROOT` and `DOTNET_ROOT` point to. It is not the behavior that I would expect.

marcpopMSFT commented on accepted/2025/sdk-selection-improvements.md:
```diff
@@ -0,0 +1,176 @@
+## **Proposal: DOTNET\_SDK\_ROOT**
+
+Introduce a new environment variable:  
+`DOTNET_SDK_ROOT`  
```
That's a good question. We'd need to understand what scenarios customers have where they are setting DOTNET_ROOT today. In particular, cases where they are setting it to a different location than their SDK and on a box where they are using their SDK.

Do we have any idea how common that is? I assume we don't have a lot of info for this so we'd be a bit blind (which is what makes me nervous about doing that). I'd be ok with one value though fwiw.

marcpopMSFT committed:
Add some clarification based on PR feedback. Remove the bolded headers

marcpopMSFT committed:
Updating the document for a new redirect proposal and adding a section on previous alternatives that were considered.

If this is still too much risk for the runtime team, we'll have to proceed with the guidance to customers modifying their profile/path.

jkotas commented on accepted/2025/sdk-selection-improvements.md:
```diff
@@ -0,0 +1,269 @@
+
+*   These environment variables only affect SDK loading behavior when running `dotnet` commands that require SDK resolution.
+*   Direct execution of applications via `dotnet foo.dll` is **not affected** to avoid impacting runtime behavior.
+*   If the host is launched from the source path specified in `DOTNET_ROOT_REDIRECT_SOURCES`, SDK resolution will use the path specified in `DOTNET_ROOT_REDIRECT_TARGET`.
```
I dislike CLI tools that change the behavior based on the path they are launched from. It means e.g. copying/linking the tool to some other place or launching it via an alternative path produces different results...

marcpopMSFT commented:
This proposal was strictly focused on trying to enable us to get feedback from previews on machines that had Windows Admin installs and the muxer is unique enough that it may require a unique solution. I went ahead and added some compact ecosystem context to the primary document that outlines the cli-acquisition proposal [here](https://github.com/dotnet/designs/pull/353/files).

I'm going to leave this open as we continue to have conversations but per [my comment](https://github.com/dotnet/designs/pull/350#discussion_r2604728794) above, we have a few options that are potentially more temporary in nature to unblock feedback and we'll focus on those. We'll gather feedback on usage of dotnetup previews and use that to drive how we solve the admin problem long-term.

Nirmal4G commented:
Seems to me that this is becoming too complicated just for a few edge cases, with much less benefits across the board.

As I see, there are 3 levels of installation/usage for any kind of App/SDKs.—the system level, user level and the local (repo/directory) level. Look at how Git and NuGet configs works. Can't you guys design something simpler like that?

Also, just like Git and NuGet configs, can't you introduce `dotnet.json` or `dotnet.config` to replace `global.json` as the latter seems to be too generic in nature?

marcpopMSFT commented:
Closing for now. As we get feedback on the experience, we may revisit options.
